### DISCUSSION TITLE
Use BetterNodeFinder for phpstan rule development

### DISCUSSION THREAD

**Discussion by staabm** [rectorphp/rector#6921] [Q&A] [answered]
I would love to use the BetterNodeFinder for my phpstan extension development.

atm it seems its necessary to require the whole rector codebase to make use of it... is my observation correct?

I tried requiring rector/rector, but even then I cannot use the BetterNodeFinder via constructor-injection.
Is there some config file which I need to configure so the injection container gets aware of `BetterNodeFinder`... I couldn't find a config in the rector code base which would define something related to `BetterNodeFinder`.

thx for any hints.

**Comment by TomasVotruba** [comment, marked as answer]
BetterNodeFinder depends on some mechanics related to Rector run itself.

For PHPStan rules, we use this one:
https://github.com/symplify/astral/blob/main/src/NodeFinder/SimpleNodeFinder.php

If these is some method missing in `SimpleNodeFinder`, feel free to add it/move it. Its purpose is for PHPStan rules

**Reply by staabm** [reply]
I am interessted in `findPreviousAssignToExpr`, which requires also `NodeComparator` etc.

I will try if I can get these ported over

----

```php
$sql = 'SELECT name, colour, calories
    FROM fruit
    WHERE calories < :calories AND colour = :colour';
$sth = $dbh->prepare($sql, array(PDO::ATTR_CURSOR => PDO::CURSOR_FWDONLY));
$sth->execute(array('calories' => 150, 'colour' => 'red'));
$red = $sth->fetchAll();
```

`findPreviousAssignToExpr` would give me the `$sth = $dbh ..` assignment,  based on a `Expr` of `$sth->fetchAll()` right?

**Reply by staabm** [reply]
hmm I guess just having `findFirstPrevious` in the astral package `SimpleNodeFinder` would be cool.

this would be less code to move, but still reqiures a `AttributeKey::CURRENT_STATEMENT` which does not yet exist in symplify.
